README
Proc::ProcessTable, version 0.44

Please use rt.cpan.org to submit bugs and patches.

STATUS
======
This is BETA software; it seems to work, but use at your own risk :) 

Currently works on darwin, nonstop-ux, Cygwin on Windows, linux,
solaris, aix, hpux, freebsd, irix, dec_osf, bsdi, netbsd, unixware 7.x,
SunOS and openbsd. Please see the "README.osname" files for details on
individual os implementations. Please see the file PORTING if you are
interested in making it work on something else. Please see the file
TODO for a list of issues that need to be addressed (and send me
patches!).

Please note that the Cygwin/windows port requires the Cygwin
environment to work (available from http://cygwin.com/), and that the
code for the port is derived from Cygwin code and is therefore covered
by the Cygwin license (http://cygwin.com/licensing.html).

Multithread support is now available for Solaris; please see
README.solaris for info. It may work under other OS's as well; please
let me know if it does.

Comments, bug reports, patches and especially ports are greatly
appreciated. If you want to submit a patch, *please* use standard
context-diff format; if you're submitting a port, a tarball of the new
files is great.

DESCRIPTION
===========
This module is a first crack at providing a consistent interface to
Unix (and maybe other multitasking OS's) process table information.
The impetus for this came about with my frustration at having to parse
the output of various systems' ps commands to check whether specific
processes were running on different boxes at a larged mixed Unix site.
The output format of ps was different on each OS, and sometimes
changed with each new release of an OS. Also, running a ps subprocess
from within a perl or shell script and parsing the output was not a
very efficient or aesthetic way to do things.

With this module, you can do things like this:

	# kill memory pigs 
	use Proc::ProcessTable;

	$t = new Proc::ProcessTable;
        foreach $p ( @{$t->table} ){	
	  if( $p->pctmem > 95 ){
	    $p->kill(9);
          }		
        }

There is another short example in the file "example.pl" in the
distribution. For a more elaborate example (in German), see
<http://www.linux-magazin.de/ausgabe.1999.02/Proc/proc.html>.
<shameless plug> If you can't read German, try my other module,
WWW::Babelfish!</shameless plug> 
	
There are also two contributed modules: a module called Proc::Killall
contributed by Aaron Sherman to kill all processes whose command-lines
match a given pattern, and a module called Proc::Killfam by Stephen
Lidie to kill a list of processes and their children. These modules
are installed along with Proc::ProcessTable. Pod documentation is
included in both of them.


INSTALLATION
============
This module needs the File::Find and Storable modules in order to
work. File::Find is generally included with perl distributions;
Storable is available from CPAN. 

After unpacking the tar file, do:

        perl Makefile.PL 
        make
	make test
        make install

There is embedded POD documentation in ProcessTable.pm and
Process/Process.pm.

ACKNOWLEDGEMENTS
================
Thanks to the many people who have sent in ports and patches. Without
them this module would be impossible to support on so many platforms.
Patches are noted in the Changes file.

David Paquet <David.Paquet@cnes.fr>		AIX port
Mike Romberg <romberg@fsl.noaa.gov>		HPUX port
Slaven Rezic <eserte@cs.tu-berlin.de>		FreeBSD port
W. Phillip Moore <wpm@ms.com>			IRIX port
Peter ? <hooft@natlab.research.philips.com>	IRIX version patch
Bernhard Schmalhofer <Bernhard.Schmalhofer@gmx.de>	dec_osf port
Sean Eskins <sean@gilasoft.com>				bsdi port
Peter Reich <pr@alles.prima.de>				netbsd port
Aaron Sherman <ajs@ajs.com>		Proc::Killall module
Steve Lidie <sol0@Lehigh.EDU>		Killfam.pm module
Martin Lucina <mato@catalyst.net.nz>	Unixware 7.x port
Shawn Clifford <shawn.a.clifford@lmco.com> SunOS port
J Robert Ray <jrray@jrray.org>		Windows (Cygwin) port.
Tom Wyant <twyant3@comcast.net>		Darwin port.
Mike Steinert <mike.steinert@motorola.com>
					Nonstop-UX port.
<bsd@openbsd.rutgers.edu>		Openbsd port.

Please note that Bernard Schmalhofer is no longer able to provide
support for the dec_osf port.
			
COPYRIGHT
=========
Copyright (c) 1998-2008 Daniel J. Urist. All rights reserved.
This package is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

--
Daniel J. Urist
durist@frii.com


Changes
Revision history for Perl extension Proc::ProcessTable.

0.01  Mon Jun  8 11:37:47 1998
	- original version; created by h2xs 1.18

0.03  Sun Jul 26 09:07:50 1998
	- added argument to va_end to make linux-ppc work

0.04  Fri Aug 14 22:51:31 1998
        - preliminary pre-alpha solaris support

0.05  Wed Dec  9 18:47:39 1998
	- fixed minor compilation problems for solaris. Changed rss to
	bytes on linux for consistency with solaris; also added "time"
	and "ctime" fields to linux for consistency. Updated
	documentation in Process.pm. Folded in AIX port by David
	Paquet <david-paquet@usa.net>.

0.06  Wed Dec  9 23:42:05 1998
	- added HPUX port by Mike Romberg <romberg@fsl.noaa.gov>.

0.07  Wed Jan  6 14:57:04 1999
	- added FreeBSD port by Slaven Rezic <eserte@cs.tu-berlin.de>.

0.08  Thu Feb  4 00:35:53 1999 
	- Fixed bug in solaris hints file that was looking for a
	version of solaris returned by "uname -r" as > 2.5, when
	"uname -r" returns the "SunOS" version of 5.5. Fixed bug in
	linux module in which only the first field of the command line
	was getting passed in cmndline attribute.

0.09   Fri Feb 19 23:23:53 1999 
	- Fixed bug in Linux.c, fopen() called without corresponding
	fclose. Added empty DESTROY subs to ProcessTable.pm and
	Process.pm to fix occasional problems with mod_perl. Changed
	all instances of sv_undef in ProcessTable.xs to PL_sv_undef to
	comply with new standard.

0.10	Sat Feb 20 11:29:35 1999
	- Added code to ProcessTable.xs to check perl patchlevel
	version and use either PL_sv_undef or sv_undef since lots of
	people are still using older versions of perl.

0.11	Sat Mar  6 19:04:39 1999
	- Fixed code that checked perl patchlevel version to check
	directly for definition of PL_sv_undef and define it as
	sv_undef if it's not there, since it's not clear what
	patchlevel we need to check for. Also fixed embarrassing and
	unnecessary call to close() a string in Solaris.c

0.12	Sun Mar 21 10:53:56 1999
	- Hopefully the final fix for the sv_undef buglet.

0.14	Fri May 14 06:30:15 1999 
	- Added port for IRIX from W. Phillip Moore <wpm@ms.com> in
	0.13, but forgot to add files to MANIFEST and edit Changes
	(Duh). W. Phillip Moore's port also included a minor patch for
	Solaris to get the page size from the system. 

0.15	Tue Jun  1 21:33:16 1999
	- Added patch from Peter ? <hooft@natlab.research.philips.com>
	for IRIX version differences. There seems to be a lot of IRIX
	out there.

0.16	Mon Aug  9 22:59:46 1999
	- Added nanosecond Solaris patch from Rolf Petter Halle
	<rph@nextel.no>. Added port for dec_osf from Bernhard
	Schmalhofer <Bernhard.Schmalhofer@fazi.de>, which included new 
	version of test code as well. Added bsdi port from Sean Eskins
	<sean@gilasoft.com>. 

0.17	Thu Aug 19 10:57:16 1999
	- Lost connection to ISP during upload to PAUSE, and PAUSE
	won't let me reload it so I have to increment the version
	number. Kind of annoying.

0.18	Thu Aug 19 10:57:16 1999
	- Fixed bug (at least for linux) in Makefile.PL that was
	causing ProcessTable.xs not to be converted to ProcessTable.c; 
	this was carelessly introduced by me with integration of the
	port to dec_osf. Added port for netbsd from Peter Reich
	<pr@alles.prima.de>.

0.19	Mon Aug 30 00:22:02 1999
	- Reinstated 'OBJECT' argument in WriteMakefile in Makefile.PL 
	since it was breaking the build on dec-osf, but this time with 
	explicit object files to look for which seems to work, at
	least under Linux. Why this should be necessary is
	unclear. Added "fname" field to HPUX (was "cmd") for
	consistency with other ports and for new test. Added very
	basic "README.hpux". Added check in ProcessTable.pm to prune
	search tree for unreadable directories when finding TTY list;
	this was causing spurious errors during build on dec_osf
	since some directories under /dev are readable only by root.

0.20	Wed Sep  8 22:27:28 1999
	- Added "fields" method to module to return list of fields
	supported for the current architecture. This is implemented in 
	a slightly kludgy way, since it needs to call the table method 
	if the table method hasn't been called before, since only the
	table method knows the names of the fields. Also changed the
	test code to dump all the fields/values available for the
	current process table, instead of just dumping pids and
	"fname" which isn't supported on all architectures (though it
	should be). Added "example.pl" file which contains the same
	code. 

0.21	Fri Sep 10 12:32:13 1999
	- Bug fix; new "fields" method was crapping out on everything
	except Linux-Intel.

0.22	Tue Oct  5 07:47:54 1999 
	- Bug fix for Solaris; nanosecond process times are only
	available on 2.6 (and above?); this was causing the module to
	dump core on earlier systems. Also, size and rss appear to be
	in kilobytes, not pages in 2.6 and above (they are in pages in 
	the struct prpsinfo in < 2.6, but also available directly as
	bytes). 

0.23	Mon Nov  1 08:32:54 1999
	- Added patch from Dave Alden <alden@math.ohio-state.edu> for
	solaris "nice" and "onproc" fields.

0.24	Thu Jan 20 08:00:30 2000
	- Added patches from Noel Paul <NoelP@mincom.com> for dec_osf
	pctcpu and pctmem, and for IRIX for pctmem.

0.25	Thu Feb  3 11:20:27 2000
	- Added patch from F. Buirey <f.buirey@saprr.fr> for AIX SMP.

0.26	Fri Feb 11 16:43:48 2000
	- Added Proc::Killall from Aaraon Sherman <ajs@ajs.com>.

0.27	Thu Jun 29 08:27:23 2000
	- Added patch for converting jiffies to microseconds in linux
	from Philip Gwyn <fil@pied.nu>. Added patch from Slaven Rezic
	<eserte@cs.tu-berlin.de> for FreeBSD and to make Storable
	optional. 

0.28	Mon Aug 14 16:31:43 MYT 2000
	- Patch from Slaven Rezic <eserte@cs.tu-berlin.de> to make
       test script -w clean.

0.29    Tue Jan  9 07:36:13 EST 2001
       - Patch from Wolfgang Friebel <Wolfgang.Friebel@desy.de> for
       os/HPUX.c to make the time method return useful values. Patch
       from Adrian Phillips <tandem@dnmi.no> to fix AIX cmndline

os/Linux.h
#include <linux/limits.h>
#include <fcntl.h>
#include <asm/param.h>
#include <stdlib.h>
#include <stdio.h>
#include <ctype.h>
#include <time.h>
#include <unistd.h>
#include <dirent.h>
#include <string.h>
#include <sys/resource.h>
#include <sys/stat.h>
#include <sys/vfs.h>
#include <sys/types.h>

#ifndef ARG_MAX
#define ARG_MAX sysconf(_SC_ARG_MAX)
#endif

/****************************************/
/* Process state strings that we return */
/****************************************/
#define SLEEP  "sleep"
#define WAIT   "wait"       
#define RUN    "run"
#define IDLE   "idle"
#define ZOMBIE "defunct"
#define STOP   "stop"
#define UWAIT  "uwait"      

/***********************************************************/
/* Some global variables we only need to get once          */
/***********************************************************/
long Btime;
unsigned Sysmem;

/***********************************************************/
/* stuff stored in /proc/xxx/stat it would be nice if this */
/* was in a system header file somewhere                   */
/***********************************************************/
struct procstat {
  int pid;
  char comm[FILENAME_MAX];
  char state;
  int ppid;
  int pgrp;
  int session;
  int tty;
  int tpgid;
  unsigned flags;
  unsigned minflt;
  unsigned cminflt;
  unsigned majflt;
  unsigned cmajflt;
  long long utime;
  long long stime;
  long long cutime;
  long long cstime;
  int counter;
  int priority;
  unsigned timeout;
  unsigned itrealvalue;
  unsigned long  starttime;
  unsigned long vsize;
  unsigned long rss;
  unsigned rlim;
  unsigned startcode;
  unsigned endcode;
  unsigned startstack;
  unsigned kstkesp;
  unsigned kstkeip;
  int signal;
  int blocked;
  int sigignore;
  int sigcatch;
  unsigned wchan;
};

/* We need to pass in a cap for ignore, lower for store on object */
/* We can just lc these! */
static char Defaultformat[] = "IIIIIIIIIIIIIJJJJJJPPISLSSSSSIIIIIIS";

/* Mapping of field to type */
static char* Fields[] = {
  "uid",
#define F_UID 0 

  "gid",
#define F_GID 1

  "pid",
#define F_PID 2

  "ppid",
#define F_PPID 3

  "pgrp",
#define F_PGRP 4

  "sess",
#define F_SESS 5

  "priority",
#define F_PRIORITY 6

  "ttynum",
#define F_TTYNUM 7

  "flags",
#define F_FLAGS 8

  "minflt",
#define F_MINFLT 9

  "cminflt",
#define F_CMINFLT 10

  "majflt",
#define F_MAJFLT 11

  "cmajflt",
#define F_CMAJFLT 12

  "utime",
#define F_UTIME 13

  "stime",
#define F_STIME 14

  "cutime",
#define F_CUTIME 15

  "cstime",
#define F_CSTIME 16

  "time",
#define F_TIME 17

  "ctime",
#define F_CTIME 18

  "size",
#define F_SIZE 19

  "rss",
#define F_RSS 20

  "wchan",
#define F_WCHAN 21

  "fname",
#define F_FNAME 22

  "start",
#define F_START 23

  "pctcpu",
#define F_PCTCPU 24

  "state",
#define F_STATE 25

  "pctmem",
#define F_PCTMEM 26

  "cmndline",
#define F_CMNDLINE 27

  "exec",
#define F_EXEC 28

  "euid",
#define F_EUID 29

  "suid", 
#define F_SUID 30

  "fuid",
#define F_FUID 31

  "egid", 
#define F_EGID 32

  "sgid", 
#define F_SGID 33

  "fgid",
#define F_FGID 34

  "cwd"
#define F_CWD 35

#define F_LASTFIELD 35
};
